### PR TITLE
Statepoint file loading refactor and CAPI function

### PR DESCRIPTION
**Pull request by pshriwise**
# Description

This PR contains a slight refactor of the statepoint loading so it can more cleanly be exposed as part of the CAPI. 

It is also a continuation of #2390 in that the scenario where a statepoint is loaded with `n_batches` equal to `settings::n_max_batches` is softened to a warning. This allows previously executed statepoint files can be loaded into an initialized `openmc.lib` instance to recover tally information. An additional check for the current batch vs the max number of batches is added to `openmc_next_batch` to ensure that no particles will be transported in the case that the warning condition is triggered. 

I think this results in reasonable behavior in the case that a user feeds a statepoint file from a complete OpenMC run back into the `openmc` executable without modification of the batches in the settings file: OpenMC will initialize, load the statepoint file, produce the warning message as part of the statepoint file load, run zero batches, and finalize. Importantly, it will not overwrite the existing statepoint file in this case either. It's somewhat unfortunate that a user will have to wait for problem initialization to get this result, but this was true before these changes as well.

We're using this as part of a project with Cardinal to load reaction rate tallies into a separately initialized OpenMC problem to leverage the depletion module for the Bateman eq. solve before sending updated material compositions back into Cardinal.

I went ahead and added a section on simulation restarts as I didn't see any information on it outside of the restart flag passed to the executable.

The restart regression test needed some updating after this to use a fixture that allowed for a check that the correct warning is present. 

# Checklist

- [x] I have performed a self-review of my own code
- [x] I have run [clang-format](https://docs.openmc.org/en/latest/devguide/styleguide.html#automatic-formatting) (version 15) on any C++ source files (if applicable)
~- [ ] I have followed the [style guidelines](https://docs.openmc.org/en/latest/devguide/styleguide.html#python) for Python source files (if applicable)~
- [x] I have made corresponding changes to the documentation (if applicable)
- [x] I have ~added~ modified tests that prove my fix is effective or that my feature works (if applicable)

### CODE REVIEW

**Comment by pshriwise** [comment]
Update: to comply with expected behavior of `openmc.lib`'s ability to iterate on batches beyond what is specified in the `settings.xml`, I moved the batches check into the `openmc_run` function so the early exit only applies when calling the `openmc` executable or `openmc.lib.run`.

File: src/state_point.cpp [hunk 1]
```diff
@@ -421,9 +430,10 @@ void load_state_point()
   read_dataset(file_id, "current_batch", simulation::restart_batch);
 
   if (simulation::restart_batch >= settings::n_max_batches) {
-    fatal_error(fmt::format(
-      "The number of batches specified for simulation ({}) is smaller"
-      " than the number of batches in the restart statepoint file ({})",
+    warning(fmt::format(
```
Comment (gridley) [inline src/state_point.cpp:433]: With that `if` statement you added earlier, doesn't that mean we will just show this warning then exit? Just making sure I understand correctly.

File: tests/regression_tests/statepoint_restart/test.py [hunk 2]
```diff
@@ -61,24 +61,37 @@ def test_statepoint_restart():
...
-        # update the number of batches and run again
+        model.settings.batches = 6
+        # ensure we capture output only from the next run
+        capsys.readouterr()
+        sp_file = model.run(export_model_xml=False, restart_file=sp_file)
+        # indicates that a new statepoint file was not created
+        assert sp_file is None
+
+        output = capsys.readouterr().out
+        output = re.sub(' +', ' ', output.replace('\n',' '))
```
Comment (gridley) [inline tests/regression_tests/statepoint_restart/test.py:85]: Why did you need to do this? I don't understand it, sorry. It seems like you can just look for the strings below without this line.

**Review by gridley — CHANGES_REQUESTED** [review]
very minor requests!

File: openmc/lib/core.py [hunk 3]
```diff
@@ -95,6 +95,11 @@ class _SourceSite(Structure):
 _dll.openmc_statepoint_write.argtypes = [c_char_p, POINTER(c_bool)]
 _dll.openmc_statepoint_write.restype = c_int
 _dll.openmc_statepoint_write.errcheck = _error_handler
+_dll.openmc_statepoint_load.argtypes = [c_char_p]
+_dll.openmc_statepoint_load.restype = c_int
+_dll.openmc_statepoint_load.errcheck = _error_handler
+_dll.openmc_statepoint_write.restype = c_int
```
[the remark this answers is not in the document]
Comment (pshriwise) [reply openmc/lib/core.py:101]: I only see one place in `openmc_statepoint_write` with a return value and it returns `0`, so I think in practice failures would come from a `fatal_error` call and this is more for completeness. As discussed in https://github.com/openmc-dev/openmc/issues/1778, this probably isn't ideal.

File: src/simulation.cpp [hunk 4]
```diff
@@ -51,9 +51,15 @@
 
 int openmc_run()
 {
+  using namespace openmc;
```
[the remark this answers is not in the document]
Comment (pshriwise) [reply src/simulation.cpp:54]: Fair point. That's inconsistent with how other variables are referenced in our CAPI functions (for the most part). Updating...

File: src/state_point.cpp [hunk 5]
```diff
@@ -364,11 +364,27 @@ void restart_set_keff()
...
+  write_message(
+    fmt::format("Loading state point {}...", settings::path_statepoint_c), 5);
+  openmc_statepoint_load(settings::path_statepoint.c_str());
+}
+
+void statepoint_version_check(hid_t file_id)
+{
+  // Read revision number for state point file and make sure it matches with
+  // current version
+  array<int, 2> array;
```
[the remark this answers is not in the document]
Comment (pshriwise) [reply src/state_point.cpp:376]: Fixed!

File: src/state_point.cpp (hunk 1 above)
Comment (pshriwise) [reply src/state_point.cpp:433]: That's correct, but with this change OpenMC will no longer won't exit here. When using OpenMC as a library, this is important b/c a statepoint file (and its results) can be loaded and accessed without killing the library even if the number of batches in the file relative to those set in the simulation isn't appropriate for a restart run.

File: tests/regression_tests/statepoint_restart/test.py (hunk 2 above)
Comment (pshriwise) [reply tests/regression_tests/statepoint_restart/test.py:85]: Ah, that was a hold-over from a different set of strings I was using to perform the check. Removing now (along with the unecessary `print` in the line below).

**Review by paulromano — APPROVED** [review]
Thanks for this improvement @pshriwise!

@gridley Can you let us know if you're OK with merging?
